instrumentation/opentelemetry-instrumentation-threading/src/opentelemetry/instrumentation/threading/__init__.py

```diff
@@ -0,0 +1,101 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=empty-docstring,no-value-for-parameter,no-member,no-name-in-module
+
+from concurrent.futures import thread
+import threading # pylint: disable=import-self
+from os import environ
+from typing import Collection
+from opentelemetry import context
+from wrapt import wrap_function_wrapper as _wrap
+
+
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+
+from opentelemetry.instrumentation.threading.package import _instruments
+from opentelemetry.instrumentation.threading.version import __version__
+
+from opentelemetry.trace import (
+    INVALID_SPAN,
+    INVALID_SPAN_CONTEXT,
+    get_current_span,
+    get_tracer_provider,
+    get_tracer,
+    SpanKind
+)
+
+ATTRIBUTE_THREAD_NAME = "currentthread.name"
+DEFAULT_THREAD_NAME = "thread"
+
+def _with_tracer_wrapper(func):
+    """Helper for providing tracer for wrapper functions."""
+
+    def _with_tracer(tracer):
+        def wrapper(wrapped, instance, args, kwargs):
+            return func(tracer, wrapped, instance, args, kwargs)
+
+        return wrapper
+
+    return _with_tracer
+
+def _wrap_target(ctx, target_func, tracer):
+    """Helper for providing tracer for wrapper functions."""
+    context.attach(ctx)
+    with tracer.start_as_current_span(
+        "threading.Thread.target",
+        kind=SpanKind.INTERNAL,
+    ) as span:
+        if span.is_recording():
+            span.set_attribute(ATTRIBUTE_THREAD_NAME, target_func.__name__)
+    return target_func
+
+@_with_tracer_wrapper
+def _wrap_thread(tracer, wrapped, instance, args, kwargs):
+    """Wrap `Threading.thread`"""
+
+    target_func = kwargs.get("target")
+    
+    with tracer.start_as_current_span(
+        "threading.Thread",
+        kind=SpanKind.INTERNAL,
+    ) as span:
```

Comment:
Is there a need to start span here?

Reply:
I think yes because that is why thread object would be part of the trace.

Reply:
What value does it add? And taking another closer look, I think this doesn't work when the thread instance is created outside (in your test, move the `t1 = threading.Thread(target=self.print_square, args=(10)) ` outside of with).

What you would probably need is subclassing the original thread something like this.

```patch
diff --git a/instrumentation/opentelemetry-instrumentation-threading/src/opentelemetry/instrumentation/threading/__init__.py b/instrumentation/opentelemetry-instrumentation-threading/src/opentelemetry/instrumentation/threading/__init__.py
index 4dee6610..834b77b7 100644
--- a/instrumentation/opentelemetry-instrumentation-threading/src/opentelemetry/instrumentation/threading/__init__.py
+++ b/instrumentation/opentelemetry-instrumentation-threading/src/opentelemetry/instrumentation/threading/__init__.py
@@ -14,7 +14,6 @@

 # pylint: disable=empty-docstring,no-value-for-parameter,no-member,no-name-in-module

-from concurrent.futures import thread
 import threading # pylint: disable=import-self
 from os import environ
 from typing import Collection
@@ -27,6 +26,7 @@ from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
 from opentelemetry.instrumentation.threading.package import _instruments
 from opentelemetry.instrumentation.threading.version import __version__

+from opentelemetry import trace
 from opentelemetry.trace import (
     INVALID_SPAN,
     INVALID_SPAN_CONTEXT,
@@ -79,9 +79,25 @@ def _wrap_thread(tracer, wrapped, instance, args, kwargs):
             span.set_attribute(ATTRIBUTE_THREAD_NAME, wrapped.__name__)
         return wrapped(*args, **kwargs)

+
+class _InstrumentedThread(threading.Thread):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._parent_span = None
+
+    def start(self):
+        self._parent_span = get_current_span()
+        super().start()
+
+    def run(self):
+        parent_span = self._parent_span or get_current_span()
+        trace.set_span_in_context(parent_span)
+        super().run()
+
+
 class ThreadingInstrumentor(BaseInstrumentor):  # pylint: disable=empty-docstring

-    start_func = None
+    original_threadcls = threading.Thread

     def instrumentation_dependencies(self) -> Collection[str]:
         return _instruments
@@ -92,9 +108,9 @@ class ThreadingInstrumentor(BaseInstrumentor):  # pylint: disable=empty-docstrin
         tracer_provider = kwargs.get("tracer_provider", None) or get_tracer_provider()

         tracer = get_tracer(__name__, __version__, tracer_provider)
-        _wrap(threading, "Thread", _wrap_thread(tracer))
+        threading.Thread = _InstrumentedThread
+        _InstrumentedThread._tracer = tracer


     def _uninstrument(self, **kwargs):
-        unwrap(threading, "Thread")
-
\ No newline at end of file
+        threading.Thread = self.original_threadcls
```

Reply:
I tried without the `with` statement
```
def test_thread_with_root(self):
       # with self.tracer.start_as_current_span("root"):
        t1 = threading.Thread(target=self.print_square, args=(10))
        t1.start()
        t1.join()

        spans = self.memory_exporter.get_finished_spans()
        print(spans[0].__dict__)
        print(spans[1].__dict__)

```
Following is the output, looks like it is even working without the with statement
```
{'_name': 'threading.Thread.target', '_context': SpanContext(trace_id=0x875c4a7bcd5c05c4f8b251824595debb, span_id=0xeebc77da57e9d56f, trace_flags=0x01, trace_state=[], is_remote=False), '_kind': <SpanKind.INTERNAL: 0>, '_instrumentation_info': InstrumentationInfo(opentelemetry.instrumentation.threading, 0.37b0.dev, ), '_instrumentation_scope': InstrumentationScope(opentelemetry.instrumentation.threading, 0.37b0.dev, ), '_parent': SpanContext(trace_id=0x875c4a7bcd5c05c4f8b251824595debb, span_id=0x52a399ad98e1bf38, trace_flags=0x01, trace_state=[], is_remote=False), '_start_time': 1675701957063421160, '_end_time': 1675701957063458875, '_attributes': BoundedAttributes({'currenttarget.name': 'print_square'}, maxlen=128), '_events': BoundedList([], maxlen=128), '_links': BoundedList([], maxlen=128), '_resource': <opentelemetry.sdk.resources.Resource object at 0x7f66957454f0>, '_status': <opentelemetry.trace.status.Status object at 0x7f6695734d00>}
{'_name': 'threading.Thread', '_context': SpanContext(trace_id=0x875c4a7bcd5c05c4f8b251824595debb, span_id=0x52a399ad98e1bf38, trace_flags=0x01, trace_state=[], is_remote=False), '_kind': <SpanKind.INTERNAL: 0>, '_instrumentation_info': InstrumentationInfo(opentelemetry.instrumentation.threading, 0.37b0.dev, ), '_instrumentation_scope': InstrumentationScope(opentelemetry.instrumentation.threading, 0.37b0.dev, ), '_parent': None, '_start_time': 1675701957063325198, '_end_time': 1675701957063561961, '_attributes': BoundedAttributes({'currentthread.name': 'Thread'}, maxlen=128), '_events': BoundedList([], maxlen=128), '_links': BoundedList([], maxlen=128), '_resource': <opentelemetry.sdk.resources.Resource object at 0x7f66957454f0>, '_status': <opentelemetry.trace.status.Status object at 0x7f6695734d00>}

```
But, brainstorming over your suggestion now.